tools/cargo-compare/src/check.rs

```diff
@@ -0,0 +1,96 @@
+
+use crate::common::GuppyCargoCommon;
+use crate::GlobalContext;
+use anyhow::{Error, Result};
+use diffus::{edit, Diffable};
+use guppy::PackageId;
+use guppy_cmdlib::CargoMetadataOptions;
+use proptest::prelude::*;
+use proptest::test_runner::{Config, TestCaseResult, TestError, TestRunner};
+use std::collections::{BTreeMap, BTreeSet};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use structopt::StructOpt;
+
+/// Options for cargo/guppy comparisons.
+#[derive(Debug, StructOpt)]
+pub struct CheckOpts {
+    /// Number of randomly generated diff operations to run
+    #[structopt(long, default_value = "256")]
+    pub cases: u32,
+    #[structopt(flatten)]
+    pub metadata: CargoMetadataOptions,
+}
+
+impl CheckOpts {
+    /// Executes this command.
+    pub fn exec(self, ctx: &GlobalContext) -> Result<()> {
+        let strat = GuppyCargoCommon::strategy(&self.metadata, ctx.graph);
+
+        let mut testrunner = TestRunner::new(Config {
+            cases: self.cases,
+            ..Config::default()
+        });
+
+        // print a message after every n tests
+        let test_count = AtomicUsize::new(0);
+        let print_every = 5;
+
+        testrunner
+            .run(&strat, |common| {
+                let cargo_map = common
+                    .resolve_cargo(ctx)
+                    .map_err(|e| TestCaseError::fail(e.to_string()))?;
+                let guppy_map = common
+                    .resolve_guppy(ctx)
+                    .map_err(|e| TestCaseError::fail(e.to_string()))?;
+
+                assert_diff(
+                    "target diff, guppy -> cargo",
+                    &guppy_map.target_map,
+                    &cargo_map.target_map,
+                )?;
+
+                assert_diff(
+                    "host diff, guppy -> cargo",
+                    &guppy_map.host_map,
+                    &cargo_map.host_map,
+                )?;
+
+                if test_count.fetch_add(1, Ordering::SeqCst) >= print_every {
+                    println!("finished runnning {} tests", print_every);
+                    test_count.store(0, Ordering::SeqCst);
+                }
+
+                Ok(())
+            })
+            .map_err(|e| match e {
+                TestError::Abort(e) => {
+                    Error::msg(format!("Aborted cargo/guppy diff check, {:?}", e))
+                }
+                TestError::Fail(e, v) => {
+                    Error::msg(format!("Failed cargo/guppy diff check {:?}\n{:?}", e, v))
+                }
+            })
+    }
+}
+
+fn assert_diff(
+    descr: &'static str,
+    a: &BTreeMap<PackageId, BTreeSet<String>>,
+    b: &BTreeMap<PackageId, BTreeSet<String>>,
+) -> TestCaseResult {
+    if let edit::Edit::Change(diff) = a.diff(&b) {
+        for (pkg_id, diff) in diff {
+            prop_assert!(
+                !diff.is_copy(),
+                "unexpected diff ({}) {}: {:?}",
+                descr,
+                pkg_id,
+                diff
```

Comment:
I've added a `FeatureDiff` struct to `diff.rs` that would be useful here, could you try switching to that? Thanks!